### DISCUSSION TITLE
How to provide additional *.ko

### DISCUSSION THREAD

**Discussion by masahir0y** [ipxe/ipxe#502] [General]
I want to provide an additional module when booting the kernel.

I found this page:

  https://ipxe.org/cmd/imgfetch

The line

  initrd /modules/2.6.31/e1000.ko /lib/modules/e1000.ko"

... is presumably what I wanted to achieve.

I do not know how it works, but if my understanding is correct,
/lib/modules/e1000.ko will appear in the initrd, correct?

My ipxe script is like this:

    #!ipxe

    kernel tftp://192.168.122.1/vmlinuz
    initrd tftp://192.168.122.1/initrd.img
    initrd tftp://192.168.122.1/ice.ko /lib/modules/ice.ko

    imgstat

    boot vmlinuz console=ttyS0,115200 initrd=initrd.img rdinit=/bin/sh

The result is like this.
The boot booted and successfully entered the shell
(I gave rdinit=/bin/sh)

iPXE 1.21.1+ (g2265) -- Open Source Network Boot Firmware -- http://ipxe.org
Features: DNS HTTP iSCSI TFTP SRP AoE EFI Menu
tftp://192.168.122.1/boot.ipxe... ok
tftp://192.168.122.1/vmlinuz... ok
tftp://192.168.122.1/initrd.img... ok
tftp://192.168.122.1/ice.ko... ok
vmlinuz : 6769256 bytes [EFI] [SELECTED]
initrd.img : 55129656 bytes
ice.ko : 31390248 bytes "/lib/modules/ice.ko"
[    0.000000] Initializing cgroup subsys cpuset
[    0.000000] Initializing cgroup subsys cpu
[    0.000000] Initializing cgroup subsys cpuacct

  (snip kernel logs)
  

sh-4.2# ls /lib/modules/
3.10.0-1160.el7.x86_64  keys
sh-4.2# 

I checked the /lib/modules/, but ice.ko does not exist there.

I am missing something?

**Comment by mcb30** [comment]
As described on https://ipxe.org/cmd/imgfetch, you will need to use `initrd=initrd.magic` in order to provide the Linux kernel with the injected files.  You can do this by editing your script to have the line:
```
boot vmlinuz console=ttyS0,115200 initrd=initrd.magic rdinit=/bin/sh
```

(On a separate note, consider using a protocol such as HTTP instead of TFTP - it will be substantially faster.)

**Reply by masahir0y** [reply]
Ah!

I missed to read the document carefully.

Now it works. Much appreciated!

**Reply by masahir0y** [reply]
BTW, if I use a released version of ipxe,
initrd=initrd.magic results in an error
(Failed to open file: initrd.magic).

I am guessing this feature is very new
(commit e5f02551)
so I need to use the git version.

My system is based on UEFI, not legacy BIOS.

**Reply by NiKiZe** [reply]
Yes it is new, but available in the builds at boot.ipxe.org

Please always consider git master as stable

**Reply by sbrudenell** [reply]
> Please always consider git master as stable

That is very surprising, and more than a little scary.

Could you document this fact, possibly at https://ipxe.org/download#source_code ?

**Reply by NiKiZe** [reply]
I think it already is plenty documented. As you don't have any other download links for sources other than the git repo I'm not sure how this is an issue?

**Reply by sbrudenell** [reply]
I (and OP?) got confused by the existence of tagged versions. Github presents these as releases, and indeed many projects manage releases just by pushing tags. The most recent tag of ipxe is "only" 18 months old, so they seemed relevant.

If the version tags aren't relevant anymore, perhaps they should be deleted. If this isn't realistic, perhaps the situation should be documented.

I expected ipxe to have stable releases anyway, since bootloaders are critical code. It's scary to test a bleeding-edge release on a reboot. If it fails I need a rescue disk, and I won't even know what ipxe commit to use as "last known good", unless I can rollback my entire configuration which isn't always realistic. I can smoke test with qemu but it only gets me so far.

**Reply by NiKiZe** [reply]
So again current release is at boot.ipxe.org (built every hour from master)
